### PR TITLE
Return empty result when a group by gets optimized to a timeseries query

### PR DESCRIPTION
**Pull request by LakshSingla**
### Description
Related to #11188 

The above mentioned PR allowed timeseries queries to return a default result, when queries of type: `select count(*) from table where dim1="_not_present_dim_"` were executed. Before the PR, it returned no row, after the PR, it would return a row with value of `count(*)` as 0 (as expected by SQL standards of different dbs).

In `Grouping#applyProject`, we can sometimes perform optimization of a groupBy query to a timeseries query if possible (when the keys of the groupBy are constants, as generated by automated tools). For example, in `select count(*) from table where dim1="_present_dim_" group by "dummy_key"`, the groupBy clause can be removed. However, in the case when the filter doesn't return anything, i.e. `select count(*) from table where dim1="_not_present_dim_" group by "dummy_key"`, the behavior of general databases would be to return nothing, while druid (due to above change) returns an empty row. This PR aims to fix this divergence of behavior.

Example cases:
1. `select count(*) from table where dim1="_not_present_dim_" group by "dummy_key"`.
_CURRENT_: Returns a row with count(*) = 0
_EXPECTED_: Return no row

2. `select 'A', dim1 from foo where m1 = 123123 and dim1 = '_not_present_again_' group by dim1`
_CURRENT_: Returns a row with ('A', 'wat')
_EXPECTED_: Return no row

To do this, a boolean `droppedDimensionsWhileApplyingProject` has been added to `Grouping` which is true whenever we make changes to the original shape with optimization. Hence if a timeseries query has a grouping with this set to true, we set `skipEmptyBuckets=true` in the query context (i.e. donot return any row).
<hr>

##### Key changed/added classes in this PR
 * `Grouping.java`
 * `DruidQuery#toTimeseriesQuery`

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: sql/src/main/java/org/apache/druid/sql/calcite/rel/Grouping.java [hunk 1]
```diff
@@ -70,12 +71,25 @@ private Grouping(
...
+      final RowSignature outputRowSignature,
+      final boolean droppedDimensionsWhileApplyingProject
+  )
   {
     this.dimensions = ImmutableList.copyOf(dimensions);
     this.subtotals = Preconditions.checkNotNull(subtotals, "subtotals");
     this.aggregations = ImmutableList.copyOf(aggregations);
     this.havingFilter = havingFilter;
     this.outputRowSignature = Preconditions.checkNotNull(outputRowSignature, "outputRowSignature");
+    this.droppedDimensionsWhileApplyingProject = droppedDimensionsWhileApplyingProject;
```
Comment (abhishekagarwal87) [inline sql/src/main/java/org/apache/druid/sql/calcite/rel/Grouping.java:92]: can we rename it to something else e.g. hadGroupingDimensions or groupDimensionsDropped or something? basically, we need not add `whileApplyingProject` here

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 2]
```diff
@@ -13342,4 +13340,34 @@ public void testCommonVirtualExpressionWithDifferentValueType() throws Exception
         ImmutableList.of()
     );
   }
+
+  // When optimization in Grouping#applyProject is applied, and it reduces a Group By query to a timeseries, we
+  // want it to return empty bucket if no row matches
+  @Test
+  public void testReturnEmptyRowWhenGroupByIsConvertedToTimeseries() throws Exception
```
Comment (abhishekagarwal87) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:13347]: can you add few more queries such as one that groups on a dummy dimension

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java [hunk 3]
```diff
@@ -13342,4 +13340,34 @@ public void testCommonVirtualExpressionWithDifferentValueType() throws Exception
...
+  {
+    skipVectorize();
+    testQuery(
+        "SELECT 'A', dim1 from foo WHERE m1 = 50 AND dim1 = 'wat' GROUP BY dim1",
```
Comment (abhishekagarwal87) [inline sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:13351]: btw I noticed that it runs as a group-by query if we remove the `m1 = 50` clause. do you know why?

**Review by abhishekagarwal87 — COMMENTED** [review]
Minor comments. LGTM otherwise.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 3 above)
Comment (LakshSingla) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java:13351]: When removing the `m1=50` clause, the `dim1` is not getting reduced to `wat` literal in the Calcite planner phase, so the optimization in `Grouping.java` to eliminate the literals is not getting applied. 
I checked the place where this reduction is happening, and it's in the `ProjectMergeRule` of Calcite. When there's a single project, then `ProjectMergeRule` is not getting invoked, which is causing this reduction to literal.

File: sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java (hunk 2 above)
Comment (LakshSingla) [reply sql/src/test/java/org/apache/druid/sql/calcite/CalciteQueryTest.java]: Added a few more test cases, which show the consequences of optimization in different cases.

**Review by abhishekagarwal87 — APPROVED** [review]

**Comment by abhishekagarwal87** [comment]
Merged since test failures are unrelated.

**Comment by jihoonson** [comment]
@abhishekagarwal87 I feel our test is too flaky recently. Can you file an issue for flaky tests in this PR, so that we can promote and fix it? If we have one already, please link it to here.

**Comment by abhishekagarwal87** [comment]
Sure thing @jihoonson

**Comment by jihoonson** [comment]
Thank you!

**Pull request merged**